### PR TITLE
Photon: Force only Photon URLs to HTTPS

### LINKED ISSUE #6073: `jetpack_photon_domain` filter value is forced https
Given a filter:

```
add_filter( 'jetpack_photon_domain', function( $domain ){
	return 'http://photon.dev';
});
```

Since 0fa10d4288b383429114926f5327f676366bb345, `http://photon.dev` is forced to be `https` when I'd expect it to remain `http`.

If a custom domain is provided through the `jetpack_photon_domain` filter, it doesn't make sense to force a scheme.

### PR DESCRIPTION
**Pull request by aduth**
Fixes #6073

#### Changes proposed in this Pull Request:

This pull request seeks to resolve an issue where providing one's own Photon domain via the `jetpack_photon_domain` would not have its scheme respected and instead forced to HTTPS, regardless of whether the custom domain is capable of supporting it.

The changes here force only those URLs we detect to be Photon to HTTPS. Combined with previous changes in 0fa10d4, this should ensure all Photon URLs use HTTPS while still respecting custom domains provided through the filter.

Further, it resolves an issue where if a custom domain is provided for Photon that is already "network_path" and a valid scheme is provided, that scheme should be reflected in the return value.

#### Testing instructions:

Verify newly added unit tests are passing:

```
WP_DEVELOP_DIR=/path/to/wordpress-develop/tests/phpunit phpunit tests/php/test_functions.photon.php 
```

Replace `WP_DEVELOP_DIR` with a path to a checkout of the [WordPress develop repository](https://wordpress.org/download/source/).

[_(See original testing instructions)_](https://gist.github.com/aduth/87e319476ae6ecf21233dc860b891d1c)

cc @danielbachhuber @jeherve

### CODE REVIEW

**Comment by danielbachhuber** [comment]
No unit tests?

**Comment by aduth** [comment]
Good call, they'd not existed 'til semi-recently. I'll add cases based on testing instructions.

**Comment by aduth** [comment]
Added unit tests and updated testing instructions accordingly.

I also changed behavior to not try to override the scheme of a URL which is already "Photonized". It seems as though we should respect the scheme of the Photon domain*, though I think this might need to be part of a larger set of changes to deprecate the `$scheme` parameter altogether and simply enforce the scheme of the Photon domain.

_*_ In other words, I think `jetpack_photon_url( 'http://i0.wp.com/example.com/img.jpg' )` should probably return `https://i0.wp.com/example.com/img.jpg`, but with the implementation here will preserve the original scheme.

**Comment by georgestephanis** [comment]
@aduth The unit tests added seem to have a couple failures in PHP 5.2/5.3 -- 

test_photon_url_filter_network_path_photonized_http
test_photon_url_filter_network_path_photonized_https
test_photon_url_filter_network_path_photonized_to_https

If you need me to, I can dig into what's causing the failures, just let me know.

**Comment by aduth** [comment]
@georgestephanis I may or may not have time to take a look this week, but I suspect it could be an existing issue with URL parsing related to this PHP Changelog note from 5.4.7:

>Fixed host recognition when scheme is omitted and a leading component separator is present.

http://php.net/manual/en/function.parse-url.php#refsect1-function.parse-url-changelog

At this line:

https://github.com/Automattic/jetpack/blob/df97b9d/functions.photon.php#L101

**Comment by georgestephanis** [comment]
I'll put a reminder on it and try to hit it later this week if you don't beat me to it.

Alternately, if anyone finds this still open after next week, please hunt me down and beat me with a whisky bottle until I fix it.

**Comment by aduth** [comment]
In 2e53352f08bb839a6486fc9df3173284d8239c09, added `jetpack_photon_parse_url` as a wrapper to `parse_url` which prepends an assumed scheme for schema-less URLs. Might make sense as something more generic than how it is currently (i.e. not specific to Photon).
